src/main/java/com/mathworks/ci/MatlabScriptBuilder.java

```diff
@@ -0,0 +1,146 @@
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.StaplerRequest;
+import hudson.EnvVars;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.Launcher.ProcStarter;
+import hudson.model.AbstractProject;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.Builder;
+import jenkins.tasks.SimpleBuildStep;
+import net.sf.json.JSONObject;
+
+public class MatlabScriptBuilder extends Builder implements SimpleBuildStep {
+    private int buildResult;
+    private EnvVars env;
+    private MatlabReleaseInfo matlabRel;
+    private CommandConstructUtil cmdUtils;
+    private String matlabCommand;
+
+    @DataBoundConstructor
+    public MatlabScriptBuilder() {
+
+
+    }
+
+
+    // Getter and Setters to access local members
+
+
+    @DataBoundSetter
+    public void setMatlabCommand(String matlabCommand) {
+        this.matlabCommand = matlabCommand;
+    }
+
+    public String getMatlabCommand() {
+        return this.matlabCommand;
+    }
+
+    private String getCommand() {
+        return this.env == null ? getMatlabCommand() : this.env.expand(getMatlabCommand());
+    }
+
+    private void setEnv(EnvVars env) {
+        this.env = env;
+    }
+
+    @Symbol("RunMatlabCommand")
+    @Extension
+    public static class MatlabScriptDescriptor extends BuildStepDescriptor<Builder> {
+
+
+        MatlabReleaseInfo rel;
+
+
+        // Overridden Method used to show the text under build dropdown
+        @Override
+        public String getDisplayName() {
+            return Message.getValue("Builder.script.builder.display.name");
+        }
+
+        @Override
+        public boolean configure(StaplerRequest req, JSONObject formData) throws FormException {
+            save();
+            return super.configure(req, formData);
+        }
+
+        /*
+         * This is to identify which project type in jenkins this should be applicable.(non-Javadoc)
+         * 
+         * @see hudson.tasks.BuildStepDescriptor#isApplicable(java.lang.Class)
+         * 
+         * if it returns true then this build step will be applicable for all project type.
+         */
+        @Override
+        public boolean isApplicable(
+                @SuppressWarnings("rawtypes") Class<? extends AbstractProject> jobtype) {
+            return true;
+        }
+    }
+
+    @Override
+    public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace,
+            @Nonnull Launcher launcher, @Nonnull TaskListener listener)
+            throws InterruptedException, IOException {
+
+        // Set the environment variable specific to the this build
+        setEnv(build.getEnvironment(listener));
+
+        // Create command util for command constrcution.
+        String matlabRoot = this.env.get("matlabroot");
+        cmdUtils = new CommandConstructUtil(launcher, matlabRoot);

```

Comment:
My initial implementation was to call MATLAB dumbly. with a caveat . due to Java ProcessLaunchers limitations (known currently to me) the default shell process would not get hold of the Environment variables that we set in the wrapper class. it could be only accessed by the child process. hence to make it work i had to invoke the command similar to below
/bin/bash -c "matlab -batch" (on Linux)
cmd.exe /C matlab -batch (on Windows)
both work fine with -batch option however when it comes to -r option the command construction cribs about parenthesis & spaces. It also requires to escape all the "(" , ")" & Single quotes which looks like below.
![image](https://user-images.githubusercontent.com/47204011/74906691-c10ab580-53d7-11ea-841a-b08079c081d5.png)

Currently, I am investigating this issue and checking how we could resolve it and make the command look as clean as possible (Without any escape characters &  prefixing shell program executables. this issue was causing delays to out review process hence implemented it in this way for time being. I shall raise new PR once i get to convincing solution for the same.

Reply:
Perhaps the use of `run_matlab_command` would help. Not sure if you're ready to switch over yet but it is available at:

https://ssd.mathworks.com/supportfiles/ci/run-matlab-command/v0/run-matlab-command.zip

You could invoke it like you're doing with `-batch` and it should take care of things for you:

Linux:
`/bin/bash -c "run_matlab_command.sh runMatlabTests(...)"`

Windows:
`cmd.exe /C run_matlab_command.bat runMatlabTests(...)`

Reply:
If it is easy to incorporate I like the idea of using run_matlab_command, but we'd probably need to include it with the Jenkins Plugin as a build step, grabbing the "released" version from the location Mark references above. 

I also think that it would be good to incorporate the script generator, but perhaps that is something that can wait until later? It seems like if starting with run_matlab_command helps resolve this problem and it doesn't consistute a large change in scope then it ay be worth doing now.

Reply:
I think its doable. lot of my code will be reduced with this. working on it to adopt the changes . need bit of testing. i will update the review request once done.

Reply:
I think copying the zip file, putting it into resources, and then including it in the jar/hpi is the right approach, but I would say we should not check in those files but instead do it as part of the build.

Reply:
* I have made some changes to pom file to resolve the dependancy run time and package it.
i am making dure that the zip file is unpacked and copied under reources folder and then it will be packaged in .jar  
* Adopted Mark's changes.
* both builder classes are changed accordingly.
* Run tests works perfectly fine on both Windows / Linux 
* Run command works perfectly in Windows.

* We still have an issue with /bin/bash -c option when  it comes to running Custom command s. i still need to escape for commands like disp() and seems to be an issue still to identify commands with parenthesis and single quotes even if we escape them as shown below.
![image](https://user-images.githubusercontent.com/47204011/75245498-d591e800-57f3-11ea-9370-32f0a834e103.png)

![image](https://user-images.githubusercontent.com/47204011/75245551-f35f4d00-57f3-11ea-9467-3b70e029a010.png)

Reply:
This one is resolved now. since the restriction of using /bin/bash was due to availability of PATH variable changes only to child process. run_matlab_command script itself acts as child process and PATH will be available to the script. so we now no need to escape the path and use /bin/bash 

![image](https://user-images.githubusercontent.com/47204011/75342794-228dc100-58bd-11ea-86ed-3e81d1b0a07e.png)